### DISCUSSION TITLE
How to fix: "PL2303HXA PHASED OUT SINCE 2012. PLEASE CONTACT YOUR SUPPLIER"

### DISCUSSION THREAD

**Discussion by drmason789** [xdream-biking/trixter-xdream-bike#8] [Q&A] [answered]
You've installed the PL2303 driver that came with the X-Dream software in Windows 10 so that it can communicate with the X-Dream bike. But you get this in Device Manager:

![image](https://user-images.githubusercontent.com/29954900/143041310-3154e015-a34d-426e-8972-c9d90bc1e9ab.png)

What to do?

**Comment by drmason789** [comment, marked as answer]
To fix this, you need to restore the version of the driver that was supplied with the X-Dream software. To do this, right click the device and select Update driver:
 
![image](https://user-images.githubusercontent.com/29954900/143041536-1c868ca1-cef5-4225-8e0f-d0f7a5a1497b.png)

Then “Browse my computer for drivers”:
 
![image](https://user-images.githubusercontent.com/29954900/143041576-8928f7b2-f579-47d7-a1db-3d1ebfcde9d8.png)


Then “Let me pick from a list of available drivers on my computer”
 
![image](https://user-images.githubusercontent.com/29954900/143041599-da3718a4-dd81-4b0a-bad2-c6847d172029.png)

Select version a pre-2012 version from the list and click Next.
 
![image](https://user-images.githubusercontent.com/29954900/143041630-e3380cff-8207-4691-b461-3a57f8894f0e.png)


Hopefully this will be successful. Click the close button.
 
![image](https://user-images.githubusercontent.com/29954900/143041690-43760913-5296-457a-b593-3c7acbc763e0.png)

If it does not indicate that it is using COM2
 
![image](https://user-images.githubusercontent.com/29954900/143041720-65a3931e-5f74-4bf3-9a1a-e84ce5ab9cd2.png)

you will need to configure it as before. 

At this point, maybe after a restart, you should be able to turn on the X-Dream bike and use the “XDream Test” console application to show that the X-Dream bike and computer are communicating.

**Comment by drmason789** [comment]
Unfortunately, Windows 10, before very long, will restore the newer driver.

![image](https://user-images.githubusercontent.com/29954900/143041877-0942ece9-e443-4cf5-8c95-189656708f8b.png)

A solution is to force the use of an older driver. To stop Windows 10 from updating the driver for a specific device, you can apply some registry settings. In Windows 10 Professional, Enterprise and Educational editions, you can use gpedit.msc to edit group policy, and in all of those AND Home Edition you can make the resulting changes in the registry directly.

Copy the following text into a text file on the computer’s desktop and rename the file to pl2303.reg:

```
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Microsoft\Windows\DeviceInstall\Restrictions]
"DenyDeviceIDs"=dword:00000001
"DenyDeviceIDsRetroactive"=dword:00000000


[HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Microsoft\Windows\DeviceInstall\Restrictions\DenyDeviceIDs]
"1"="USB\\VID_067B&PID_2303&REV_0300"
"2"="USB\\VID_067B&PID_2303"
```

On a home computer, this is likely to be alright. However, there are some things you need to check.

1. Check the Hardware IDs
2. Check for a pre-existing Device Installation Restriction Policy
3. Confirm the working PL2303 driver is installed
4. Merge the pl2303.reg file

**1 Check the Hardware IDs**

Are the hardware IDs (USB\\VID_067B&PID_2303&REV_0300 and USB\\VID_067B&PID_2303) correct for this particular X-Dream bike?

Locate the Prolific USB to Serial Comm Port in Device Manager, right click it, and select Properties. 
Go to the Details tab, and select the Hardware IDs property. The values should match the list in the reg file you’ve created. If they don’t, click them in the Value list with the right mouse button, select and copy and paste them into the pl2303.reg file in your text editor.

![image](https://user-images.githubusercontent.com/29954900/143042316-ee7d7c55-2a5b-4bde-b265-fd851f66928d.png)

**2 Check for a Pre-existing Device Installation Restriction Policy**

If you simply merge the pl2303.reg file into the registry, you may overwrite an existing policy that has been set up by some other process, e.g. by Group Policy. You need to avoid this.

To determine if there is an existing policy in place, do the following:

Open regedit (WindowsKey-R, then enter regedit)

![image](https://user-images.githubusercontent.com/29954900/143042451-b4d730e3-6401-44ae-b7ed-1d0bfbef6d4e.png)

Navigate as far as you can go to this key:
 HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Microsoft\Windows\DeviceInstall\Restrictions\DenyDeviceIDs
It most likely isn’t there, and you will only be able to go as far as HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Microsoft\Windows

![image](https://user-images.githubusercontent.com/29954900/143042497-c5a95607-cba8-4a0e-a328-3753db010c41.png)

If this key is not present, there’s no need to proceed any further.

If the key is present, you’re going to need to interpret what’s already there and decide what to do.
If it appears as above, i.e. with the correct hardware IDs listed under DenyDeviceIDs, and under Restrictions:
DenyDeviceIDs set to 1
DenyDeviceIDsRetroactive set to 0

![image](https://user-images.githubusercontent.com/29954900/143042558-9da98249-a516-4000-bdd7-1bc04ca71314.png)

then it’s already set up and you don’t need to do anything other than confirm the working PL2303 driver is installed (next section).

Otherwise, you’re going to have to merge the changes into the existing set. 
You must have DenyDeviceIDs set to 1 for the policy to work.
You must have DenyDeviceIDsRetroactive set to 0 or not present to prevent Windows from removing the existing device with the older, working driver.

If these are already set to something else, you will be interfering with existing configuration and it is up to you what you do with it.
You will need to renumber the hardware ids in your pl2303.reg file so they don’t overwrite the devices that are already listed under the DenyDeviceIDs key.

Make a backup of the DeviceInstall key by right clicking and selecting Export. You might need this later.

**3 Confirm the working PL2303 driver is installed**

In Device Manager, it should look like this:

![image](https://user-images.githubusercontent.com/29954900/143042759-4aa26de5-8518-4ba9-bb17-2dfcbf1b472f.png)
 
and not this:
 
![image](https://user-images.githubusercontent.com/29954900/143042736-86409507-fca9-4194-b55f-a19ed72f51b5.png)

If it’s wrong, follow the instructions in the previous post to make it right.

**4 Merge the PL2303.reg file**

Right click the pl2303.reg file and select Merge. Click yes to approve the change:
![image](https://user-images.githubusercontent.com/29954900/143042861-16db1eb0-ac66-477a-aa26-c554e13e75ca.png)

And Yes to the warning:



![image](https://user-images.githubusercontent.com/29954900/143042911-365f6c8f-f82b-47de-8870-daa09f62b9ad.png)



If you then navigate to HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Microsoft\Windows\DeviceInstall\Restrictions you should see the new items. 
Firstly, the DenyDeviceIDs name under the Restrictions key. Its value of 1 indicates that denying the listed device IDs is enabled. You can change this to 0 to turn it off. 

![image](https://user-images.githubusercontent.com/29954900/143042956-44353dcf-89f8-4759-84ae-5935ad879688.png)

The name value pair DenyDeviceIDsRetroactive indicates if the policy should be applied to devices that are already installed. This must be disabled (0), or the X-Dream bike won’t even work with the right device driver:

![image](https://user-images.githubusercontent.com/29954900/143043024-89c58ea3-8181-4561-b519-59be286695b6.png)

The X-Dream bike’s internal PL2303’s hardware IDs are listed under the DenyDeviceIDs key in name/value pairs of increasing integer number (1,2,..)

![image](https://user-images.githubusercontent.com/29954900/143043073-e7272e7d-9957-42d1-9de7-3562df34f94b.png)

Configured this way, Windows 10 should not try to update the driver. This effectively tells Windows, if one of these devices is plugged in, if it’s not installed already, don’t install it (i.e. do recognise that it’s there, but don’t connect it to the operating system by way of a driver). 

This way, because the PL2303 with the old driver is already installed, Windows 10 accepts it and doesn’t (shouldn’t) try to adjust it.

If something changes the DenyDeviceIDsRetroactive value to 1, Windows will apply the policy to the existing device installation as well, and the device will stop working.

**Caveats**

This configuration can fail for various reasons, e.g.
-	PL2303 (X-Dream bike) is plugged into a different USB socket, or into a hub plugged into the same socket.
-	The configured registry entries are reconfigured by another process, e.g. Group Policy
-	A user manually updating the driver

**Comment by chungcu201** [comment]
Suddenly my PL2303 driver was removed by Windows 10. Has anyone have the same issue?
Tried uninstalling/installing the driver multiple times.
Tried messing with the Group Policy editor too. Nothing work.

![Screenshot 2022-04-03 222216](https://user-images.githubusercontent.com/79088552/161427848-d98b6a9b-28d2-4437-bef1-feded241fce3.jpg)
![Screenshot 2022-04-03 222130](https://user-images.githubusercontent.com/79088552/161427850-5b010529-7305-4aa1-a042-8e2faef23674.jpg)

**Reply by drmason789** [reply]
Hi,

I don't know how the system could have got this way without manual intervention by a user or by an admin somewhere doing something with group policy. If the latter, you may have a problem that you need to sort out with them. For example, if you are using a company computer for your X-Dream bike, they might not be very keen on you having this deprecated driver installed. If it's your computer, we should be able to fix this.

I have created a new discussion for this problem: #88

**Comment by drmason789** [comment]
In release 1.4.0, I've added a new feature to provide advice on the group policy situation.

Click the "Driver" button:

![image](https://user-images.githubusercontent.com/29954900/162634984-cb0c46f9-8ed4-436f-a611-b751901e84c1.png)

This will read the registry to make an assessment about the Group Policy state. 

![image](https://user-images.githubusercontent.com/29954900/162635036-7cefcf55-da64-46ef-ace5-4c361a33dc28.png)

If the current state of the registry is due to edits that have taken place since the last system reboot, it may be necessary to restart the machine before the opinion offered by this utility is accurate.

Although this utility reads the registry, it cannot write to it. In the future I may add the facility to reconfigure this by calling the Windows REG.EXE utility with a generated registry script, but running it would prompt the user for admin approval.
